### DISCUSSION TITLE
Lokinet Browser? Lokinet Messenger?

### DISCUSSION THREAD

**Discussion by Destroyinator69420** [oxen-io/lokinet#1717] [Ideas]
Could be based on Firefox and Thunderbird. Security hardening + liblokinet integration

**Comment by ghost** [comment]
Current PGP standart doesn not encrypt subject lines, I wouldn't recomend a messaging service based on popular email services
Saying that their data would be anonymous would be lying to them
You are better forking the Tutanota apps or any other one that doesn't support SMTP or POP and then integrate lokinet with their service
Email sucks btw

**Comment by ghost** [comment]
A browser for lokinet should be based on the Tor browser, so both shared similar fingerprint protections
I seriously doubt that a fork is necessary, maybe just some little tweaks
Maybe something to put on here as a documentation

**Reply by majestrate** [reply]
Mozilla and all of its software products do not have a future given the current downward trend in their user and developer count.

**Reply by peepo5** [reply]
What does have a future? Chromium? But then you would have to develop your own entirely new browser to keep people hidden. If firefox use kept going down, tor would still remain popular because of its use cases. A good suggestion would be to have a tor spin which would support onion sites and would support loki sites too. It would need some sort of passthrough for .loki domains and that is about it.

**Reply by majestrate** [reply]
the web doesn't really have a rosy future if you desire privacy or security.

solution: stop using web technology.

**Reply by peepo5** [reply]
? what ? do you want me to go outside and talk to grass?

**Reply by peepo5** [reply]
How are you meant to view .loki sites then?

**Reply by majestrate** [reply]
the web has been bought and paid for by google long ago, the standards body, the implementations and most of the services are under google's full control.

**Reply by majestrate** [reply]
lokinet isn't just for web browsers, it caries any traffic that can fit into an IP packet.

**Reply by peepo5** [reply]
Your personal view on the web are not really part of what I am discussing.
You are a maintainer for loki, so how would you want to view loki sites?

**Reply by majestrate** [reply]
I just use whatever web browser I have on hand.

**Reply by peepo5** [reply]
.loki sites can fingerprint you through the browser that is why the concern is there. Tor browser was made to stop fingerprinting but lokinet is not supported in tor.

**Reply by majestrate** [reply]
Fingerprinting, especially on the web, is out of scope at the moment as that is a rabbit hole that goes deeper than you can possibly imagine and we aren't staffed properly at the moment to actually make a meaningful dent in it.

**Reply by majestrate** [reply]
In general I am not in favor of perpetuating low quality technology like web tech.

**Reply by peepo5** [reply]
so what are you in favor of

**Reply by majestrate** [reply]
literally anything that isn't web based, probably. web is the bottom of the barrel.

**Reply by peepo5** [reply]
what is your definition of web

**Reply by majestrate** [reply]
any software that utilizes standards defined by the w3c for interpreting html documents with css and javascript.

**Reply by peepo5** [reply]
Why are you a maintainer then?

**Reply by majestrate** [reply]
I am lokinet's author and i can say for sure that it is not limited to just web traffic. It is so so much more than that.

**Reply by jagerman** [reply]
That's about like asking a maintainer of postfix (an SMTP server) why they are maintainers if they hate webmail (because most SMTP traffic is carrying mail that came from/goes to webmail).

**Reply by peepo5** [reply]
okay :<

**Reply by peepo5** [reply]
but still i think the idea of a dedicated browser would help with privacy

**Reply by majestrate** [reply]
it absolutely would, I just dont have time to maintain such a massive undertaking, feel free to take the initiative as I am sure there are a very large amount of people that want such a thing.

**Reply by jagerman** [reply]
We've had it on the roadmap for "at some future point" for a long time, but as @majestrate says, that's quite a large undertaking.

**Reply by peepo5** [reply]
I agree but somebody actually has to work on it.. I wonder if the lokinet team will actually consider the idea, otherwise it would have to be a community made project. I am really in favor of this being a thing if the loki team are serious about privacy. Even with lokinet exit nodes enabled, sites can use common tracking methods to deanonymize you, and without a browser like tor (which is hardened), lokinet's users are more vulnerable to deanonymization while using the web. Anyhow, I hope it is considered (tor fork with passthrough).

**Reply by majestrate** [reply]
> > We've had it on the roadmap for "at some future point" for a long time, but as @majestrate says, that's quite a large undertaking. -- @jagerman ([#1717 (reply in thread)](https://github.com/oxen-io/lokinet/discussions/1717#discussioncomment-1439327))
> 
> See [#1821 (reply in thread)](https://github.com/oxen-io/lokinet/discussions/1821#discussioncomment-1792780) i disagree it being a large undertaking.. Basically just fork a patch from tor browser that integrates the UI and sandboxing on top of icecat.

it does sound like an easy task, it may be initially to create a naive patch set. the real effort is maintaining the patch set and updating from upstream. that is not a job i want to have. If i had to maintain such I would rather maintain a patch set with brave as an upstream as they seem to know what they are doing far more than who ever is asleep at the wheel at mozilla.

**Reply by peepo5** [reply]
well mozilla do not push manifest 3 which makes ublock's life a pain unlike google

**Reply by peepo5** [reply]
Also let me remind you that it would be an open source application - you can use the advantage of developers interested for help maintaining the project

**Reply by majestrate** [reply]
well yea, google owns the web. all of it. it is long ago bought and paid for. they own the standards, the implementations (yes firefox too), the ad infra, and large parts of the hosting. if you want to make a web browser you must play ball with google. my solution for this is to not even play ball. the web is google and uncle google is not your friend.

**Reply by peepo5** [reply]
Well for the current state of the web, which most internet services require, it is good to have an option which goes away from the centralization and tracking as much as possible.

**Reply by majestrate** [reply]
It would be great to have, yes. I don't have time allocated for such, it is not a trivial task, especially if you want to do it right. Do it right, or do it twice and I don't want to make a web browser twice. The underlying problem is that there is negative economic intensives to fight fingerprinting tech on the web. It is far far far more profitable to track and fingerprint than to fight it and until THAT changes, nothing will change on the web.

**Reply by peepo5** [reply]
there will always be privacy problems on the web but with enough effort, anonymity can be done. We should make it easier.

**Reply by peepo5** [reply]
Are you hired by the oxen org btw?

**Reply by majestrate** [reply]
> Are you hired by the oxen org btw?

I am a full time salaried developer for oxen working on lokinet and other core projects, but primarily lokinet. 
A web browser is not a small product, a privacy hardened anonymous browser even less so.

**Reply by peepo5** [reply]
I understand. I think it would be good for someone at the org to make a poll on a browser and how much it is wanted compared to other things. This could give an idea to the team of the importance.

**Reply by majestrate** [reply]
it is most certainly not low priority, it is something we want to do but our current focus is mostly on session which is why such a product is not in scope at this current moment.

**Reply by majestrate** [reply]
> Afaik he's that super sleepy guy at melbourne blockchain centre talking about kinda a big journey about kinda crypto stuff :p

I have not ever spoken at whatever place that is to my understanding but I have been there once, I think.

**Comment by Pantyhose-X** [comment]
A browser for lokinet should be based on the [agregore-browser](https://github.com/AgregoreWeb/agregore-browser) and [Beacon](https://github.com/imperviousinc/beacon) and Brave Browser.

**Reply by majestrate** [reply]
i have far more faith in the brave browser team than mozilla when it comes to being able to deliver those fingerprint hardening results a lokinet browser needs.

**Reply by ghost** [reply]
Based on https://github.com/servo/servo

**Comment by majestrate** [comment]
On Sat, 06 Aug 2022 05:06:51 -0700
uranuspucksaxophone ***@***.***> wrote:

> I think Lokinet browser should be based on ungoogle-chromium, or any
> other chromium-based browser. Firefox is much less used (fingerprint
> more unique), and I think that we should instead of relying on a
> common fingerprint, using a random one that is rotated as often as
> possible, therefore, we could install addons into it, but only open
> source addons.
> 

I would agree with the sentiment of not using firefox or a derivative
as a base. On the fingerprinting front I have seen really cool progress
in getting fingerprints under control in chromium from brave. It would
be great if these could be backported to ug chromium as that would make
it a superb starting point for this venture.

**Comment by ghost** [comment]


**Reply by majestrate** [reply]
i cannot maintain a browser, my plate is full. someone else can though.

**Reply by majestrate** [reply]
> > i cannot maintain a browser, my plate is full. someone else can though.
> 
> @majestrate, What is the job of a maintainer?

to maintain lokinet itself.

**Reply by majestrate** [reply]
it likely may include basic knowledge of programming language syntax, especially c++ javascript and any others they use. this would be for knowing how to resolve any merge conflicts that occur.

**Reply by majestrate** [reply]
> > it likely may include basic knowledge of programming language syntax, especially c++ javascript and any others they use. this would be for knowing how to resolve any merge conflicts that occur.
> 
> @majestrate Can we do a test run?

feel free. let us know how it goes.

**Reply by majestrate** [reply]
> @majestrate sorry for the spam, but, where do I start?

i have no idea.

**Reply by mx5kevin** [reply]
This is the best way to implement them.

https://github.com/oxen-io/lokinet/discussions/1591

**Reply by Ashthetik** [reply]
> i cannot maintain a browser, my plate is full. someone else can though.

If possible, I could mockup a browser extension for typical user browsers (chromium and Firefox) while we (the community) wait for someone who's able to contribute to a community-led browser 

I'm aware of Google's distasteful sentiment around certain utilities, but it may suffice in the meantime to extend Loki's accessibility
